compiler: introduce an IR for the code generators

zerbina opened the pull request:
## Summary

Add the `CgNode` intermediate representation, change `astgen` (which is
renamed to `cgirgen`) to output it instead of `PNode`, and update all
code generators to use `CgNode`. In order to keep the changes required
for the transition low, the differences of the new IR compared to
`PNode` are kept minimal.

The intent is to have an IR that can be evolved independently from sem
and the macro API. Only `PNode` is replaced so far, but both `PType`
and `PSym` are planned to also get a dedicated version for use by the
code generators. In addition, `PNode` can now be evolved without the
code generators having to be considered.

## Details

The core of the changes is the introduction of `CgNode`. It, for now,
also uses a `ref`-tree-based data-representation and is very similar,
in both structure and naming, to the `PNode`-subset previously used by
the code generators, but with some small simplifications/renamings
already applied where it makes sense.

**Naming differences:**
- `nkNimNodeLit` -> `cnkAstLit`
- `nkCurly` -> `cnkSetConstr`
- `nkBracket` -> `cnkArrayConstr`
- `nkClosure` -> `cnkClosureConstr`
- `nkHiddenDeref` -> `cnkDerefView`
- `nkHiddentStdConv` -> `cnkHiddenConv`
- `nkDiscardStmt` -> `cnkVoidStmt`
- `nkExprColonExpr` -> `cnkBinding`

**Structural differences:**
- the `while` statement is replaced with the `repeat` statement, since
  conditional loops (i.e. `while` statements) don't exist after the MIR
  phase. A `cnkRepeatStmt` node has a single sub-node, which is the
  loop's body
- there are no multi-branch `if` statements, since those don't exist
  during and after the MIR phase. A `cnkIfStmt` has the same structure
  as an `nkElifBranch`
- a `cnkReturnStmt` node has no sub nodes
- instead of `emit` being represented via a pragma statement, a
  dedicated statement (`cnkEmitStmt`) is used
- `cnkPragmaStmt`s don't have sub-nodes, but instead directly store the
  pragma's name
- `cnkObjConstr` doesn't have an extra type slot in the first position 
- there is only a single literal-node kind for the `int`, `uint`,
  `float`, and `string` types
- there are no dedicated nodes for `of` and `else` in `case` statements,
  both use `cnkBranch`
- there are no `nkVarSection|nkLetSection` + `nkIdentDefs` counterparts.
  Instead, definitions use the `cnkDef` node

Many more simplifications are possible, but they are left for future
PRs.

### Producing the IR

`astgen` is changed to produce `CgNode` instead of `PNode` IR. The
`canUseView` and `flattenExpr` procedures for `PNode` are copied and
adjusted for `CgNode`.

For literals, which are transported through the MIR phase as `PNode`s,
translation-to`CgNode` logic is added (`translateLit`). For integer-like
literals, whether the output node is an `cnkIntLit` or `cnkUIntLit` node
is decided based on the input node's *type*, not on its node *kind* --
this helps catching issues with incorrectly typed integer literals. For
`float32` literals, `translateLit` already narrows the value.

Since the module's name now doesn't apply anymore, it is changed to
`cgirnode`. In addition, the module is moved to the `backend` directory
and `generateAST` is renamed to `generateIR` -- the module's
documentation is also updated.

### Code generation

All three code generators and their surroundings are adjusted to operate
on `CgNode` instead of `PNode`, which for the most part means replacing
occurrences of `PNode` with `CgNode` and adjusting for the structural
difference listed above. Unrelated refactorings and clean-ups are kept
to a minimum.

There are two sources of `PNode`s still reaching into the code
generators: constants and type AST. When generating the code for
constants, their AST is first translated to `CgNode` IR before being
passed on to normal code generation, while for the type AST case,
special-purpose routines that still use `PNode` are added.

Routines still used by the code generators that are only available for
`PNode` are copied into the new `compat` module and adjusted to use
`CgNode`. This is meant to be an interim solution, and they are planned
to be phased out and removed again in the future.

Both the C and JavaScript code generation orchestrators passed the
`PNode` body of the main procedure directly to the code generator. This
doesn't work anymore, and it is replaced with canonicalizing (which
produces a `CgNode` tree) the body first.

zerbina committed:
WIP: introduce an IR for the code generators

zerbina committed:
progress; the first bootstrap iteration works again

zerbina committed:
vm: separate argument processing from `opcExpandToAst`

The operation now expects its operands to be `NimNode`s already. A new
instruction (`opcDataToAst`) is introduced for creating the AST
representation of VM data.

While also simplifying the VM a bit, the main reason behind the change is
to not having to provide the full AST of the template call expression,
as doing so is not possible when the code generator no longer operates
on `PNode` AST.

zerbina commented:
Attempting to derive the first revision from the MIR was a mistake, as too much changes to the code generators are required for that to work. I'm going to revert all changes and start over, but now with a focus on keeping both the overall changeset and development of the new IR to a minimum.

However, there is still some decoupling left do to before the work here can resume. This includes things like making dynlib handling part of the unified backend processing pipeline, or replacing/removing dependencies on routines that are not part of the code generators and use `PNode` (e.g., `dfa.aliases`, `TNodeTable`, etc.).

zerbina committed:
start over; revert all changes so far

zerbina committed:
Merge branch 'devel' into introduce-codegen-ir

zerbina committed:
ccgexprs: remove node flag usage

The `nfAllFieldsSet` flag stopped reaching the code generator with the
introduction of the MIR, meaning that the condition always evaluates to
'true'.

zerbina committed:
add the code-generator IR

`astgen` is adjusted to produce `CgNode` instead of `PNode`. For this,
multiple `PNode` -> `CgNode` translation procedures had to be
introduced and `canUseView` + `flattenExpr` duplicated and adjusted for
`CgNode`. The general processing logic stays the same.

The module's document comment is also adjusted and an outdated mention
of "sections" (they are called "regions" in the MIR) fixed.

`astgen` as the name doesn't make much sense anymore and is going to be
changed to something more fitting.

zerbina committed:
vmaux: decouple `findMatchingBranch` from `PNode`

Instead of an integer-literal node, the procedure now accepts the value
directly.

zerbina committed:
canonicalize the main procedure's body

The C and JS code-gen orchestrators were passing the AST produced for
the main procedure directly to the code generators. This is no longer
going to work once the code generators don't work with `PNode` anymore,
so `canonicalize` is now used on the AST.

zerbina committed:
cgirutils: use `CgNode`

zerbina committed:
vmgen: decouple the emit procedures from `PNode`

The instruction-emission procedure now accept a `TLineInfo` as input
directly, instead of, unnecessarily, requiring a `PNode`. Wrappers that
still use `PNode` are added for convenience.

zerbina committed:
mirbridge/backend: use `CgNode`

`canonicalize` and `generateAST` now return `CgNode` trees. For debug
rendering, a `treeRepr` procedure for `CgNode` is added to the
`cgirutils` module.

zerbina committed:
ast_types: remove the `codegenExprNodeKinds` set

It's obsolete now.

zerbina committed:
update the code generators

All three code generators now use the `CgNode` IR. The changes to the
modules are kept minimal in order to make review easier.

As an additional way to keep the amount of changes smaller, the `compat`
module is introduced.

zerbina committed:
debugutils: add a `frameMsg` overload for `CgNode`

zerbina committed:
rename `astgen` to `cgirgen`

In addition, the module is moved to the `backend` directory.

zerbina committed:
astgen: rename `generateAST` to `generateIR`

saem commented:
i really like all the node simplification and rework, one question about it, did you already consider naming `cnkVoidStmt` as `cnkVoidExpr`? Asking because I read the node as `void`ing an expression.

zerbina commented:
> i really like all the node simplification and rework, one question about it, did you already consider naming `cnkVoidStmt` as `cnkVoidExpr`? Asking because I read the node as `void`ing an expression.

Hm, no, I didn't consider it. My thinking was that while it voids (discards) a value, it itself acts as a statement (returns no value), so, for consistency with the other statements, used the `Stmt` suffix.

saem commented on compiler/backend/cgir.nim:
```diff
@@ -0,0 +1,203 @@
+
+    cnkAsgn       ## a = b
+    cnkFastAsgn   ## fast assign b to a
+    # future direction: have ``cnkAsgn`` mean "assign without implying any
```
I wonder if a better distinction is assign vs initialize?

For context I'm simply broadly remarking upon nomenclature.

zerbina replied:
That's a good question. Everything related to assignment is very fuzzy at the moment, mainly because of "assignment" having different meaning depending on the used code generator (they don't operate on the same language level with regards to assignments).

Broadly speaking, `cnkAsgn` combines both "assign" (copy to non-empty destination) *and* "initialize" (copy to empty destination), while `cnkFastAsgn` means "create a shallow copy". Whether combining "assign" and "initialize" is a good idea, I'm not sure (it probably isn't), but `PNode` did it, so for initial compatibility, I carried it over.

saem replied:
Sounds like we're on the same page, needs some more thinking.

saem commented on compiler/backend/cgir.nim:
```diff
@@ -0,0 +1,203 @@
+    of cnkSym:        sym*: PSym
+    of cnkPragmaStmt: pragma*: TSpecialWord
+    of cnkWithItems:
+      childs*: seq[CgNode]
```
```suggestion
      kids*: seq[CgNode]
```

Shorter, grammatically "correcter" (😁), and also not needlessly gendered ("sons").

zerbina replied:
I'll change it.

saem commented on compiler/backend/cgir.nim:
```diff
@@ -0,0 +1,203 @@
+    of cnkWithItems:
+      childs*: seq[CgNode]
+
+  # future direction: move to a single-sequence-based, data-oriented design
```
♥️

saem commented on compiler/backend/jsbackend.nim:
```diff
@@ -19,6 +19,9 @@ import
     options
   ],
+  compiler/mir/[
+    mirbridge, # for `canonicalize`
```
Is this preferable as compared to a selective import?

zerbina replied:
No, a selective import is, in my opinion, preferable here. I'm unsure why I originally opted to not use one.

saem commented on compiler/backend/jsgen.nim:
```diff
@@ -704,12 +711,12 @@ proc genWhileStmt(p: PProc, n: PNode) =
   let labl = p.unique.rope
   lineF(p, "Label$1: while (true) {$n", [labl])
-  p.nested: genStmt(p, n[1])
+  p.nested: genStmt(p, n[0])
```
The index change is because of the new structure of repeat vs while?

zerbina replied:
Yep, the loops body was previously in the second slot, but now it's in the first (and only) one.

saem reviewed (approved):
this really is an incredible milestone in decoupling sem and the backends, along with backend unification! 🎉

zerbina committed:
jsbackend: use a selective import for `mirbridge`

zerbina committed:
cgir: rename `childs` to `kids`

zerbina committed:
cgirgen: work around strict-side-effect analysis bug

zerbina committed:
docs: update mentions of `astgen`

The module is named `cgirgen` now. `debug.rst` also contained outdated
mentions of `PNode` being the IR the code generators -- this is fixed
too.

zerbina commented:
Thank you for the review, @saem!

zerbina commented:
/merge
